### PR TITLE
 Dropbox synchronize folders 

### PR DESCRIPTION
**Pull request by anisg**
following up on #7286,

this feature allows you to synchronize the folder tree from dropbox when pulling files. in order to use it you go to plugins > cloud storage > long press on your dropbox cloud item > synchronize now

Screenshot 2021-02-14 at 11 35 16

Screenshot 2021-02-14 at 11 27 44
Screenshot 2021-02-14 at 16 26 46
Screenshot 2021-02-14 at 11 27 44

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/7474)

### CODE REVIEW

File: frontend/apps/cloudstorage/dropbox.lua [hunk 1]
```diff
@@ -10,9 +10,280 @@ local util = require("util")
...
 local _ = require("gettext")
+local lfs = require("libs/libkoreader-lfs")
 
 local DropBox = {}
 
+function DropBox:showFilesAndFolders(url, password)
+    return DropBoxApi:showFilesAndFoldersV2(url, password)
+end
+
+function DropBox:getRemoteFilesMap(url,password,treePath)
```
Comment (Frenzie) [inline frontend/apps/cloudstorage/dropbox.lua:21]: Our coding style says that camelCase is a function for quick identification.
```suggestion
function DropBox:getRemoteFilesMap(url, password, tree_path)
```

File: frontend/apps/cloudstorage/dropbox.lua [hunk 2]
```diff
@@ -10,9 +10,280 @@ local util = require("util")
...
+	--implem: https://stackoverflow.com/questions/1426954/split-string-in-lua
+        if sep == nil then
+                sep = "%s"
+        end
+        local t={}
+        for str in string.gmatch(s, "([^"..sep.."]+)") do
+                table.insert(t, str)
+        end
+        return t
+end
```
Comment (Frenzie) [inline frontend/apps/cloudstorage/dropbox.lua:58]: Note that we have quite a few util functions.

http://koreader.rocks/doc/modules/util.html#gsplit

File: frontend/apps/cloudstorage/dropbox.lua [hunk 3]
```diff
@@ -10,9 +10,280 @@ local util = require("util")
...
+    
+    local function _downloadRemoteFiles(path,remoteFilesDownload,item,UI)
+        local response, go_on
+        local proccessed_files = 0
+        local success_files = 0
+        local unsuccess_files = 0
+        for _, file in ipairs(remoteFilesDownload) do
+                proccessed_files = proccessed_files + 1
+                print(file.local_url)
+                local text = string.format("Downloading file (%d/%d):\n%s", proccessed_files, #remoteFilesDownload, file.text)
```
Comment (Frenzie) [inline frontend/apps/cloudstorage/dropbox.lua:257]: Translation

**Review by Frenzie — COMMENTED** [review]
Looks alright to me at a quick glance, nothing really jumps out besides needing some polishing.

**Comment by Frenzie** [comment]
Thanks, looks cool! What's really strange is that the Android Dropbox app doesn't even have that. ;-)

**Comment by vietchovui** [comment]
Coud you add Google Drive to this plugin, please?

File: frontend/apps/cloudstorage/dropbox.lua [hunk 4]
```diff
@@ -10,9 +10,280 @@ local util = require("util")
...
+        end
+	end
+    return tree
+end
+
+function DropBox:synchronize(item)
+    local function tableSlice(t, l,r)
+        if not r then r = #t end
+        local nt = {}
+        for i=l,r do
```
Comment (Frenzie) [inline frontend/apps/cloudstorage/dropbox.lua:43]:
```suggestion
        for i = l, r do
```

File: frontend/apps/cloudstorage/dropbox.lua [hunk 5]
```diff
@@ -10,9 +10,280 @@ local util = require("util")
...
+        if file.type == "folder" then
+            UI:info(T(_("Retrieving folder %1/%2…"), tree_path, file.text))
+            local subtree = DropBox:getRemoteFilesMap(file.url,password,treePath.."/"..file.text)
+            for k,subf in pairs(subtree) do
+                tree[k]=subf
+            end
+        else
+            tree[treePath.."/"..file.text] = file
+        end
+	end
```
Comment (Frenzie) [inline frontend/apps/cloudstorage/dropbox.lua:35]:
```suggestion
        end
```

File: frontend/apps/cloudstorage/dropbox.lua [hunk 6]
```diff
@@ -10,9 +10,280 @@ local util = require("util")
...
+    local function stringSplit(s, sep)
+	--implem: https://stackoverflow.com/questions/1426954/split-string-in-lua
+        if sep == nil then
+                sep = "%s"
+        end
+        local t={}
+        for str in string.gmatch(s, "([^"..sep.."]+)") do
+                table.insert(t, str)
+        end
+        return t
```
Comment (Frenzie) [inline frontend/apps/cloudstorage/dropbox.lua:57]:
```suggestion
    if sep == nil then
        sep = "%s"
    end
    local t = {}
    for str in string.gmatch(s, "([^"..sep.."]+)") do
        table.insert(t, str)
    end
    return t
```

File: frontend/apps/cloudstorage/dropbox.lua [hunk 7]
```diff
@@ -10,9 +10,280 @@ local util = require("util")
...
+    local function tableSlice(t, l,r)
+        if not r then r = #t end
+        local nt = {}
+        for i = l, r do
+            table.insert(nt,t[i])
+        end
+        return nt
+    end
+    local function stringSplit(s, sep)
+	--implem: https://stackoverflow.com/questions/1426954/split-string-in-lua
```
Comment (Frenzie) [inline frontend/apps/cloudstorage/dropbox.lua:49]:
```suggestion
    --implem: https://stackoverflow.com/questions/1426954/split-string-in-lua
```

Comment (Frenzie) [reply frontend/apps/cloudstorage/dropbox.lua:49]: Should use util.gsplit instead

**Comment by Frenzie** [comment]
It still needs polishing, but I won't do it myself

**Comment by ct2034** [comment]
What kind of polishes does this need? Who defines them? Is there a list?

**Comment by Frenzie** [comment]
See the various comments I left.

**Pull request closed without merging**